app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java

```diff
@@ -1,83 +1,311 @@
+                View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN);
+        }
+
+        // Show first start welcome wizard UI.
         setContentView(R.layout.activity_first_start);
-        Button cont = findViewById(R.id.cont);
-        cont.setOnClickListener(this);
+        mViewPager = (ViewPager) findViewById(R.id.view_pager);
+        dotsLayout = (LinearLayout) findViewById(R.id.layoutDots);
+        btnBack = (Button) findViewById(R.id.btn_back);
+        btnNext = (Button) findViewById(R.id.btn_next);
+
+        mViewPager.setOnTouchListener(new OnTouchListener() {
+            @Override
+            public boolean onTouch(View v, MotionEvent event) {
+                // Consume the event to prevent swiping through the slides.
+                v.performClick();
+                return true;
+            }
+        });
+
+        // Layouts of all welcome slides
+        layouts = new int[]{
+                R.layout.activity_firststart_slide1,
+                R.layout.activity_firststart_slide2,
+                R.layout.activity_firststart_slide3};
+
+        // Add bottom dots
+        addBottomDots(0);
+
+        // Make notification bar transparent
+        changeStatusBarColor();
+
+        mViewPagerAdapter = new ViewPagerAdapter();
+        mViewPager.setAdapter(mViewPagerAdapter);
+        mViewPager.addOnPageChangeListener(mViewPagerPageChangeListener);
+
+        btnBack.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                onBtnBackClick();
+            }
+        });
+
+        btnNext.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                onBtnNextClick();
+            }
+        });
     }
 
-    private boolean isFirstStart() {
-        return mPreferences.getBoolean("first_start", true);
+    public void onBtnBackClick() {
+        int current = getItem(-1);
+        if (current >= 0) {
+            // Move to previous slider.
+            mViewPager.setCurrentItem(current);
+            if (current == 0) {
+                btnBack.setVisibility(View.GONE);
+            }
+        }
     }
 
-    private void startApp() {
-        if (!haveStoragePermission()) {
-            requestStoragePermission();
-            /**
-             * startApp will be called in {@link #onRequestPermissionsResult()}
-             * after permission was granted.
-             */
-            return;
+    public void onBtnNextClick() {
+        // Check if we are allowed to advance to the next slide.
+        if (mViewPager.getCurrentItem() == SLIDE_POS_LOCATION_PERMISSION) {
+            // As the storage permission is a prerequisite to run syncthing, refuse to continue without it.
+            if (!haveStoragePermission()) {
+                Toast.makeText(this, R.string.toast_write_storage_permission_required,
+                        Toast.LENGTH_LONG).show();
+                this.finish();
```

Comment:
What does finish do in this case?

Reply:
I exits the app by ending the FirstStartActivity like it was the case before the PR if the user did not grant the required storage permission.

Reply:
Android development guidelines for permission request say we're not allowed to ask this multiple times on our own. That's why the button makes sense.

Reply:
We should just disable the next button if the user has not granted permissions.

Reply:
A message is already printed on the screen when exiting, so it's no crash.
Disabling the next button would leave no way to exit the foreground app window. User would then have to press home or the panic button.

Reply:
Why can't he press the back button?

Reply:
Or we could just make clicking next print the toast and not progress to the new window. Then the user can back out, or grant the permission.

Reply:
I'll think about that. Good for follow-up.